modules/security/keyvault/main.bicep

```diff
@@ -0,0 +1,100 @@
+@description('Deployment Location')
+param location string
+
+@description('Prefix of Cosmos DB Resource Name')
+param prefix string = 'kv'
+
+@description('Name of the Key Vault')
+param name string = '${prefix}-${uniqueString(resourceGroup().id)}'
+
+@description('The tenant ID where the Key Vault is deployed')
+param tenantId string = subscription().tenantId
+
+@description('For an existing Managed Identity, the Subscription Id it is located in')
+param subscriptionId string = subscription().subscriptionId
+
+@description('For an existing Managed Identity, the Resource Group it is located in')
+param resourceGroupName string = resourceGroup().name
+
+@description('Subnet ID for the Key Vault')
+param subnetID string = ''
+
+@description('List of RBAC policies to assign to the Key Vault')
+param rbacPolicies array = []
+
+@description('RBAC Role Assignments to apply to each RBAC policy')
+param roleAssignments array = [
+  '4633458b-17de-408a-b874-0445c86b69e6' // rbacSecretsReaderRole
+  'a4417e6f-fecd-4de8-b567-7b0420556985' // rbacCertificateOfficerRole
+]
+
+@allowed([ 'new', 'existing' ])
+@description('Whether to create a new Key Vault or use an existing one')
+param newOrExisting string = 'new'
+
+@description('Name of Secret to add to Key Vault')
+param secretName string = 'secret${uniqueString(resourceGroup().id)}'
+
+@secure()
+@description('Value of Secret to add to Key Vault')
+param secretValue string = ''
+
+@description('Specifies whether soft delete should be enabled for the Key Vault.')
+param enableSoftDelete bool = true
+
+@description('The number of days to retain deleted data in the Key Vault.')
+param softDeleteRetentionInDays int = 7
+
+@allowed(['standard', 'premium'])
+@description('The SKU name of the Key Vault.')
+param skuName string = 'standard'
+
+@allowed(['A', 'B'])
+@description('The SKU family of the Key Vault.')
+param skuFamily string = 'A'
+
+@description('Specifies whether RBAC authorization should be enabled for the Key Vault.')
+param enableRbacAuthorization bool = true
+
+module keyVault 'modules/vaults.bicep' = {
+  name: guid(name, 'deploy')
+  scope: resourceGroup(subscriptionId, resourceGroupName)
+  params: {
+    location: location
+    newOrExisting: newOrExisting
+    enableSoftDelete: enableSoftDelete
+    softDeleteRetentionInDays: softDeleteRetentionInDays
+    skuFamily: skuFamily
+    skuName: skuName
+    enableRbacAuthorization: enableRbacAuthorization
+    tenantId: tenantId
+    subnetID: subnetID
+    enableVNet: subnetID != ''
+  }
+}
+
+module rbacRoleAssignments 'modules/roleAssignment.bicep' = [for rbacRole in roleAssignments: {
+  name: guid(keyVault.name, rbacRole)
+  scope: resourceGroup(subscriptionId, resourceGroupName)
+  params: {
+    keyVaultName: keyVault.name
+    rbacPolicies: rbacPolicies
+    rbacRole: rbacRole
+  }
+}]
+
+module secret 'modules/secrets.bicep' = if (secretName != '' && secretValue != '') {
```

Comment:
The default value on `secretValue` should be set to the value you are checking for. Right now it auto-generates a secret name so this would never return true would it?

Reply:
I think this works? But let's review it.

We start with...
```bicep
param secretName string = 'secret${uniqueString(resourceGroup().id)}'
@secure()
param secretValue string = ''
```

So we have...
```bicep
if (secretName != '' && secretValue != '')
if (secret${uniqueString(resourceGroup().id)} != '' && '' != '')
if ( true && false )
if ( false ) // Results in section being skipped
```

Now if the user provides a value for secretValue, and leaves the default for secretName...
```bicep
if (secretName != '' && secretValue != '')
if (secret${uniqueString(resourceGroup().id)} != '' && 'Something' != '')
if ( true && true )
if ( true ) // Results in section being run
```

Now, if the user provides a secretValue, but mistakenly provides an empty string for secretName....
```bicep
if (secretName != '' && secretValue != '')
if ('' != '' && 'Something' != '')
if ( false && true )
if ( false ) // Results in section being skipped
```


I believe this works like I'm expecting?

Reply:
I poked this line, and moved the conditions into their own variable so I can more easily reuse the flag. But we should still be aligned that this works like im expecting.

Reply:
This is just me but I don't like modules auto generating names that get used in other automation workflows.

I would set the default value to an empty string to make the user provide it.

Reply:
> This is just me but I don't like modules auto generating names that get used in other automation workflows.
> 
> I would set the default value to an empty string to make the user provide it.

Since the secretName is only required when the secretValue is provided, I'm worried about if a user provides 'secretValue' but forgets to also provide 'secretName'.

Maybe I can start off just addressing this in the descriptions though, and we can see if it is a real problem or not for users.

Reply:
I changed the default value of the secretName to just ''. And I removed `secretName != ''` from the check. This way, if a user provides a secret, but forgets the name, the deployment should fail with something like invalid secret name, instead of succeeding without setting the secret.